<a href="https://colab.research.google.com/github/Nat2364/Assignment1/blob/main/Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [9]:
import zipfile
import os

zip_path = '/content/drive/My Drive/UniversityAssignmentCrimeData.zip'
output_dir = './crime_data'

os.makedirs(output_dir, exist_ok=True)

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(output_dir)

print(f"'{zip_path}' extracted to '{output_dir}'")

'/content/drive/My Drive/UniversityAssignmentCrimeData.zip' extracted to './crime_data'


In [10]:
all_files = []

for root, dirs, files in os.walk(output_dir):
    for file in files:
        if file.endswith(FILENAME_SUFFIX):
            full_path = os.path.join(root, file)
            all_files.append(full_path)

print("Identified CSV files:")
for f in all_files:
    print(f)


Identified CSV files:
./crime_data/2024-01/2024-01-west-midlands-street.csv
./crime_data/2024-06/2024-06-west-midlands-street.csv
./crime_data/2024-08/2024-08-west-midlands-street.csv
./crime_data/2024-11/2024-11-west-midlands-street.csv
./crime_data/2025-08/2025-08-west-midlands-street.csv
./crime_data/2025-11/2025-11-west-midlands-street.csv
./crime_data/2025-05/2025-05-west-midlands-street.csv
./crime_data/2025-04/2025-04-west-midlands-street.csv
./crime_data/2024-04/2024-04-west-midlands-street.csv
./crime_data/2024-03/2024-03-west-midlands-street.csv
./crime_data/2025-09/2025-09-west-midlands-street.csv
./crime_data/2025-03/2025-03-west-midlands-street.csv
./crime_data/2023-12/2023-12-west-midlands-street.csv
./crime_data/2024-07/2024-07-west-midlands-street.csv
./crime_data/2024-05/2024-05-west-midlands-street.csv
./crime_data/2025-07/2025-07-west-midlands-street.csv
./crime_data/2025-06/2025-06-west-midlands-street.csv
./crime_data/2025-01/2025-01-west-midlands-street.csv
./crim

## Load CSVs into DataFrames

### Subtask:
Iterate through the identified CSV files, reading each into a pandas DataFrame. Store these DataFrames in a list.


**Reasoning**:
To load the CSV files, I need to import the pandas library and then iterate through the list of identified file paths, reading each into a DataFrame and storing them in a list.



In [11]:
import pandas as pd

frames = []

for file_path in all_files:
    df = pd.read_csv(file_path)
    frames.append(df)

print(f"Loaded {len(frames)} DataFrames into the 'frames' list.")

Loaded 24 DataFrames into the 'frames' list.


## Concatenate DataFrames

### Subtask:
Combine all individual DataFrames from the list into one master DataFrame using `pd.concat`.


**Reasoning**:
To combine all individual DataFrames into one master DataFrame, I will use `pd.concat()` on the `frames` list and reset the index.



In [12]:
combined_df = pd.concat(frames, ignore_index=True)

print(f"Combined DataFrame shape: {combined_df.shape}")
print("First 5 rows of the combined DataFrame:")
print(combined_df.head())

Combined DataFrame shape: (668108, 12)
First 5 rows of the combined DataFrame:
                                            Crime ID    Month  \
0                                                NaN  2024-01   
1                                                NaN  2024-01   
2  08fcc46d224b1e4f4065b950fa59e5251d51106f3472cd...  2024-01   
3  9d5abe94205c0b33309388abe25b5709d77e81a42c7e54...  2024-01   
4  5ea6fbc3cd6834f56ba25852736e3e4d53a70163f2a3a3...  2024-01   

            Reported by          Falls within  Longitude   Latitude  \
0  West Midlands Police  West Midlands Police  -1.849790  52.590937   
1  West Midlands Police  West Midlands Police  -1.840582  52.598279   
2  West Midlands Police  West Midlands Police  -1.840582  52.598279   
3  West Midlands Police  West Midlands Police  -1.846312  52.593702   
4  West Midlands Police  West Midlands Police  -1.839093  52.597818   

                       Location  LSOA code        LSOA name  \
0       On or near Walsall Road  E010094

## Final Task

### Subtask:
Confirm that all 24 CSV files have been successfully loaded and merged into a single DataFrame, and confirm with the user that the data is ready for analysis.


## Summary:

### Q&A
Yes, all 24 CSV files have been successfully loaded and merged into a single DataFrame, and the data is confirmed to be ready for analysis.

### Data Analysis Key Findings
*   **File Identification:** A total of 24 CSV files matching the specified pattern (`-west-midlands-street.csv`) were successfully identified within the `./crime_data` directory.
*   **Data Loading:** All 24 identified CSV files were individually loaded into pandas DataFrames.
*   **Data Merging:** These 24 individual DataFrames were successfully concatenated into a single master DataFrame named `combined_df`.
*   **Combined DataFrame Dimensions:** The final `combined_df` contains `668,108` rows and `12` columns, indicating a large dataset ready for processing.
*   **Data Integrity Check:** The head of the `combined_df` was displayed, confirming that the data was correctly structured and preserved after the merge.

### Insights or Next Steps
*   The data is now prepared and ready for detailed exploratory data analysis, cleaning, or model building based on the crime data.
*   It would be beneficial to perform an initial data quality check, including examining data types, checking for missing values, and reviewing unique values in categorical columns.


In [13]:
combined_df

Crime ID    Month  \
0                                                     NaN  2024-01   
1                                                     NaN  2024-01   
2       08fcc46d224b1e4f4065b950fa59e5251d51106f3472cd...  2024-01   
3       9d5abe94205c0b33309388abe25b5709d77e81a42c7e54...  2024-01   
4       5ea6fbc3cd6834f56ba25852736e3e4d53a70163f2a3a3...  2024-01   
...                                                   ...      ...   
668103  3e559ffbe453cf0005fad743db1417b72d138cb7b01a41...  2024-09   
668104  b5f957c2b164d6c66ffbf81986327c3c2d617d067578e3...  2024-09   
668105  3b9505bd4b4494ffdf4f2461e4440a263de3882cba2c4e...  2024-09   
668106  19da5c4b9031a11a4efaa9079673f5db77316f69f1e933...  2024-09   
668107  c418cad2cf5bad06ea7200bce180a07895513d5e2c0b69...  2024-09   

                 Reported by          Falls within  Longitude   Latitude  \
0       West Midlands Police  West Midlands Police  -1.849790  52.590937   
1       West Midlands Police  West Midlands Police  -1.840582  52.598279   
2       West Midlands Police  West Midlands Police  -1.840582  52.598279   
3       West Midlands Police  West Midlands Police  -1.846312  52.593702   
4       West Midlands Police  West Midlands Police  -1.839093  52.597818   
...                      ...                   ...        ...        ...   
668103  West Midlands Police  West Midlands Police  -2.113522  52.576576   
668104  West Midlands Police  West Midlands Police  -2.117890  52.570935   
668105  West Midlands Police  West Midlands Police  -2.118325  52.573820   
668106  West Midlands Police  West Midlands Police  -2.119472  52.572273   
668107  West Midlands Police  West Midlands Police  -2.118325  52.573820   

                            Location  LSOA code           LSOA name  \
0            On or near Walsall Road  E01009417     Birmingham 001A   
1       On or near Badgers Bank Road  E01009418     Birmingham 001B   
2       On or near Badgers Bank Road  E01009418     Birmingham 001B   
3              On or near Hook Drive  E01009418     Birmingham 001B   
4             On or near Byron Court  E01009418     Birmingham 001B   
...                              ...        ...                 ...   
668103      On or near Furnace Drive  E01034314  Wolverhampton 035I   
668104   On or near Rowan Tree Drive  E01034315  Wolverhampton 035J   
668105         On or near Pond Grove  E01034315  Wolverhampton 035J   
668106  On or near Silver Birch Road  E01034315  Wolverhampton 035J   
668107         On or near Pond Grove  E01034315  Wolverhampton 035J   

                          Crime type  \
0              Anti-social behaviour   
1              Anti-social behaviour   
2                      Vehicle crime   
3       Violence and sexual offences   
4       Violence and sexual offences   
...                              ...   
668103  Violence and sexual offences   
668104                 Vehicle crime   
668105  Violence and sexual offences   
668106  Violence and sexual offences   
668107  Violence and sexual offences   

                                Last outcome category  Context  
0                                                 NaN      NaN  
1                                                 NaN      NaN  
2       Investigation complete; no suspect identified      NaN  
3                         Unable to prosecute suspect      NaN  
4                         Unable to prosecute suspect      NaN  
...                                               ...      ...  
668103     Action to be taken by another organisation      NaN  
668104  Investigation complete; no suspect identified      NaN  
668105                    Unable to prosecute suspect      NaN  
668106                    Unable to prosecute suspect      NaN  
668107                    Unable to prosecute suspect      NaN  

[668108 rows x 12 columns]

**Reasoning**:
To extract 'Year' and 'Month' from the existing 'Month' column, I'll first convert the 'Month' column to a datetime format using `pd.to_datetime`. Then, I'll use the `.dt.year` and `.dt.month` accessors to create new 'Year' and 'Month' columns, respectively. Finally, I'll display the head of the DataFrame to verify the new columns.

In [15]:
combined_df['Year'] = pd.to_datetime(combined_df['Month']).dt.year
combined_df['Month_num'] = pd.to_datetime(combined_df['Month']).dt.month

combined_df.head(5)

Crime ID    Month  \
0                                                NaN  2024-01   
1                                                NaN  2024-01   
2  08fcc46d224b1e4f4065b950fa59e5251d51106f3472cd...  2024-01   
3  9d5abe94205c0b33309388abe25b5709d77e81a42c7e54...  2024-01   
4  5ea6fbc3cd6834f56ba25852736e3e4d53a70163f2a3a3...  2024-01   

            Reported by          Falls within  Longitude   Latitude  \
0  West Midlands Police  West Midlands Police  -1.849790  52.590937   
1  West Midlands Police  West Midlands Police  -1.840582  52.598279   
2  West Midlands Police  West Midlands Police  -1.840582  52.598279   
3  West Midlands Police  West Midlands Police  -1.846312  52.593702   
4  West Midlands Police  West Midlands Police  -1.839093  52.597818   

                       Location  LSOA code        LSOA name  \
0       On or near Walsall Road  E01009417  Birmingham 001A   
1  On or near Badgers Bank Road  E01009418  Birmingham 001B   
2  On or near Badgers Bank Road  E01009418  Birmingham 001B   
3         On or near Hook Drive  E01009418  Birmingham 001B   
4        On or near Byron Court  E01009418  Birmingham 001B   

                     Crime type  \
0         Anti-social behaviour   
1         Anti-social behaviour   
2                 Vehicle crime   
3  Violence and sexual offences   
4  Violence and sexual offences   

                           Last outcome category  Context  Year  Month_num  
0                                            NaN      NaN  2024          1  
1                                            NaN      NaN  2024          1  
2  Investigation complete; no suspect identified      NaN  2024          1  
3                    Unable to prosecute suspect      NaN  2024          1  
4                    Unable to prosecute suspect      NaN  2024          1